src/Npgsql/NpgsqlConnector.cs

```diff
@@ -1063,7 +1064,9 @@ internal ValueTask<IBackendMessage> ReadMessage(bool async, DataRowLoadingMode d
                     try
                     {
                         // TODO: There could be room for optimization here, rather than the async call(s)
-                        ReadBuffer.Timeout = TimeSpan.FromMilliseconds(InternalCommandTimeout);
+                        ReadBuffer.Timeout = InternalCommandTimeout > 0
+                            ? TimeSpan.FromMilliseconds(InternalCommandTimeout)
+                            : Timeout.InfiniteTimeSpan;
```

Comment:
Count how many times have you repeated these lines. Makes sense to make a decision in the setter (:

Reply:
I'd just consider keeping 0 as meaning no timeout, just like before.

Reply:
I personally hate it when infinity is represented by `0` because semantically a timespan of `0` or `TimeSpan.Zero` means no time at all (immediately; pretty much the opposite of infinite) and there are a few .NET APIs where 0 means just that.
I also find it a bit dangerous if `timeout == System.Threading.Timeout.InfiniteTimeSpan` equals false in case of an infinite timeout.
While we can't change the bad choices ADO.NET made for the timeouts in the connection string variables years ago, we can choose how they are represented in our internal APIs. That's why I changed it (essentially moving the bad choices up to the ultimate public layer where I can't change them), but since 2 people seem not to share my taste, I'll revert that part.

Reply:
I agree that 0 is problematic and I do actually share your taste. But this isn't just an ADO.NET thing - [Socket.ReceiveTimeout](https://docs.microsoft.com/en-us/dotnet/api/system.net.sockets.socket.receivetimeout?view=netcore-3.1#System_Net_Sockets_Socket_ReceiveTimeout) works the same way, and IIRC this goes all the way down to Unix system calls...

Reply:
`Socket.ReceiveTimeout` is solomonic in that case:

> The default value is 0, which indicates an infinite time-out period. **Specifying -1 also indicates an infinite time-out period.**